plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/CloudLibrariesPage.java

```diff
@@ -18,106 +18,197 @@
-    
+    setSelectedLibraries(initialSelection);
+    composite.setLayout(new RowLayout(SWT.HORIZONTAL));
     setControl(composite);
   }
 
   @Override
   public boolean finish() {
-    return true;
+    final List<Library> libraries = getSelectedLibraries();
+    try {
+      if (isMavenProject) {
+        BuildPath.addMavenLibraries(project.getProject(), libraries, new NullProgressMonitor());
+      } else {
+        /*
+         * FIXME: BuildPath.addNativeLibrary() is too heavy-weight here. ClasspathContainerWizard,
+         * our wizard, is responsible for installing the classpath entry returned by getSelection(),
+         * which will perform the library resolution. We just need to save the selected libraries 
+         * so that they are resolved later.
+         */
+        Library masterLibrary =
+            BuildPath.collectLibraryFiles(project, libraries, new NullProgressMonitor());
+        newEntry = BuildPath.computeEntry(project, masterLibrary, new NullProgressMonitor());
+        BuildPath.saveLibraryList(project, libraries, new NullProgressMonitor());
+        if (newEntry == null) {
+          // container-editing only refreshes the content if new
+          BuildPath.runContainerResolverJob(project);
+        }
+      }
+      return true;
+    } catch (CoreException ex) {
+      StatusUtil.setErrorStatus(this, "Error updating container definition", ex); //$NON-NLS-1$
+      return false;
+    }
   }
 
-  @Override
-  public IClasspathEntry getSelection() {
-    // Since this class implements IClasspathContainerPageExtension2,
-    // Eclipse calls getNewContainers instead.
-    logger.log(Level.WARNING, "Unexpected call to getSelection()");
-    return null;
+  @VisibleForTesting
+  List<Library> getVisibleLibraries() {
+    List<Library> visible = new ArrayList<>();
+    for (LibrarySelectorGroup librariesSelector : librariesSelectors) {
+      visible.addAll(librariesSelector.getAvailableLibraries());
+    }
+    return visible;
   }
 
-  @Override
-  public void setSelection(IClasspathEntry containerEntry) {
-    // todo can we use the containerEntry to tick the checkboxes in the library selector group?
+  /**
+   * Return the list of selected libraries.
+   */
+  @VisibleForTesting
+  List<Library> getSelectedLibraries() {
+    List<Library> selectedLibraries = new ArrayList<>();
+    for (LibrarySelectorGroup librariesSelector : librariesSelectors) {
+      selectedLibraries.addAll(librariesSelector.getSelectedLibraries());
+    }
+    return selectedLibraries;
   }
 
-  @Override
-  public void initialize(IJavaProject project, IClasspathEntry[] currentEntries) {
-    this.project = project;
+  @VisibleForTesting
+  void setSelectedLibraries(List<Library> selectedLibraries) {
+    initialSelection = new ArrayList<>(selectedLibraries);
+    List<Library> remaining = new ArrayList<>(selectedLibraries);
+    if (!librariesSelectors.isEmpty()) {
```

Comment:
I think this check is unnecessary and actually logging a warning in that case is desirable?

Reply:
Ok.  This should never happen, and there should be warnings as part of the classpath container initializer.

Reply:
Sorry, I actually meant just removing the this `if` statement rather than removing `remaining` and the related logging, but NVM.

Now, it is really true we don't need this `if`, because the `for` loop will just exit right away.